### PR TITLE
write to file

### LINKED ISSUE #4: Implement PlpFile - a representation of the file containing plp instructions
This will be the implementation of [AsmFile](https://github.com/Progressive-Learning-Platform/PLPAssembler/blob/master/src/main/java/org/plp/isa/AsmFile.java) for file containing PLP instructions.

This should be able to read a file and similarly write to a file.

### LINKED ISSUE #5: Provide the different tokens which will constitute the PLP Program
We can take reference from [PLPTokenType](https://github.com/Progressive-Learning-Platform/PLPTool6/blob/plp_backend/src/edu/asu/plp/tool/backend/plpisa/assembler2/PLPTokenType.java). 

We should be creating the enum which will implement AsmTokenType interface.

### LINKED ISSUE #9: Load all the PlpFiles from the program Directory of a PlpProgram
When we initialize the PlpProgram, we might be loading the PlpProgram from a folder where already few asm files are present. In such a case, create PlpFile for each of these files and add it to filesInProgram during PlpProgram initialization/construction.

### LINKED ISSUE #11: For a given PlpProgram, we should be able to add existing PlpFile
This will be the implementation of assAsmFileToProgram function. 
Before adding verify if there is already a file with same name, in which case an exception needs to be thrown.

### LINKED ISSUE #12: In a PlpProgram, we should be able to create new PlpFile
This will be implementation of the createAsmFileInProgram function.
Verify if program already has file with the same name. If it has exception needs to be thrown.

### PR DESCRIPTION
**Pull request by harsha-kadekar**
With this PlpFile feature is complete. All the functions are complete.
This closes #4

---
This change is [Reviewable](https://reviewable.io/reviews/progressive-learning-platform/plpassembler/14)

### CODE REVIEW

File: src/main/java/org/plp/implementations/plpisa/PlpFile.java [hunk 1]
```diff
@@ -64,7 +72,7 @@ public String getFilePath() {
      */
     @Override
     public boolean addInstructionToFile(String instruction) {
-        return false;
+        return fileContent.add(instruction);
```
Comment (tobielf) [inline src/main/java/org/plp/implementations/plpisa/PlpFile.java:85]: instruction = "" and instruction = null;

I think the previous one can be added to the fileContent but later can not.

And we need to test the exception on the **instruction = null**

File: src/main/java/org/plp/implementations/plpisa/PlpFile.java [hunk 2]
```diff
@@ -41,8 +43,14 @@ private void validateFilePath(Path filePath) throws IOException {
      * @return true if it is successful to write to file, false if there is an error
      */
     @Override
-    public boolean writeToFile() {
-        return false;
+    public boolean writeToFile() throws IOException{
```
Comment (tobielf) [inline src/main/java/org/plp/implementations/plpisa/PlpFile.java:54]: Is it possible to mock the IOException during the Buffer Writer in our test cases?
```
@test
void writeToFileException
```

**Review by tobielf — CHANGES_REQUESTED** [review]

File: src/main/java/org/plp/implementations/plpisa/PlpFile.java [hunk 3]
```diff
@@ -41,8 +43,14 @@ private void validateFilePath(Path filePath) throws IOException {
...
+        try(BufferedWriter writer = Files.newBufferedWriter(filePath, Charset.forName("utf-8"))) {
```
Comment (themrsean) [inline src/main/java/org/plp/implementations/plpisa/PlpFile.java]: Need to check for empty file

File: src/main/java/org/plp/implementations/plpisa/PlpFile.java (hunk 1 above)
Comment (harsha-kadekar) [reply src/main/java/org/plp/implementations/plpisa/PlpFile.java:85]: Going to use [Lombok](https://objectcomputing.com/resources/publications/sett/january-2010-reducing-boilerplate-code-with-project-lombok) here with NonNull which will automatically going to throw a NullPointerException.

File: src/main/java/org/plp/implementations/plpisa/PlpFile.java (hunk 2 above)
Comment (harsha-kadekar) [reply src/main/java/org/plp/implementations/plpisa/PlpFile.java:54]: Why we want to test it?

File: src/main/java/org/plp/implementations/plpisa/PlpFile.java (hunk 3 above)
Comment (harsha-kadekar) [reply src/main/java/org/plp/implementations/plpisa/PlpFile.java]: I think empty content to the file should be good. The assumption here would be we will be using writeToFile as a ctrl+s so sometimes there is a possibility that there is no change or not started writing the program or deleting the previous content of the file and saving it.

**Review by harsha-kadekar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/progressive-learning-platform/plpassembler/14)* status: 0 of 11 files reviewed, 3 unresolved discussions (waiting on @ssohoni, @themrsean, and @tobielf)

___
*[src/main/java/org/plp/implementations/plpisa/PlpFile.java, line 75 at r1](https://reviewable.io/reviews/progressive-learning-platform/plpassembler/14#-MLajgsgvqWyisFF_uOG-r1-75:-MNp0Vlu73B7Fqfarc8f:b-cl8nva) ([raw file](https://github.com/progressive-learning-platform/plpassembler/blob/00cad80023d013d761f761571b17e5e361e72436/src/main/java/org/plp/implementations/plpisa/PlpFile.java#L75)):*
<details><summary><i>Previously, harsha-kadekar (Harsha Kadekar) wrote…</i></summary><blockquote>

Going to use [Lombok](https://objectcomputing.com/resources/publications/sett/january-2010-reducing-boilerplate-code-with-project-lombok) here with NonNull which will automatically going to throw a NullPointerException.
</blockquote></details>

I need to verify how the exception will be thrown. If it is generic then need to replace with proper exception explaining the clear reason.

**Review by ssohoni — APPROVED** [review]
Went over changes with Harsha during PLP meeting on 12/5.
Lombok errors are generic IOExceptions. Need to provide specific exceptions.
